plugins/terraform/resource_environment_test.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/quintilesims/layer0/common/models"
+	"testing"
+)
+
+func TestEnvironmentCreate_basic(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		CreateEnvironment("test-env", "m3.medium", 0, []byte("")).
+		Return(&models.Environment{EnvironmentID: "eid"}, nil)
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
+		"name": "test-env",
+	})
+
+	if err := environmentResource.Create(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentCreate_specifyOptional(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		CreateEnvironment("test-env", "m3.large", 2, []byte("user data")).
+		Return(&models.Environment{EnvironmentID: "eid"}, nil)
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
+		"name":      "test-env",
+		"size":      "m3.large",
+		"min_count": 2,
+		"user_data": "user data",
+	})
+
+	if err := environmentResource.Create(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentRead(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{})
+	d.SetId("eid")
+
+	if err := environmentResource.Read(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentUpdate(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	gomock.InOrder(
+		mockClient.EXPECT().
+			CreateEnvironment("test-env", "m3.medium", 0, []byte("")).
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&createdEnvironment, nil),
+
+		mockClient.EXPECT().
+			GetEnvironment("eid").
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&createdEnvironment, nil),
+
+		mockClient.EXPECT().
+			UpdateEnvironment("eid", 3).
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&updatedEnvironment, nil),
+
+		mockClient.EXPECT().
+			GetEnvironment("eid").
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&updatedEnvironment, nil),
+	)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d1 := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
```

Comment:
[nit] I assume you are doing the "create" steps first so that the `d.HasChange()` functions work correctly. There's probably a more explicit way to do this (maybe using [`d.Set()`](https://github.com/hashicorp/terraform/blob/master/helper/schema/resource_data.go#L150)?).